include/ncutility/BinarySerialization.h

```diff
@@ -0,0 +1,433 @@
+ */
+template<class T>
+void Deserialize(BinaryBuffer& buffer, T& out);
+
+/** @brief The maximum number of members an aggregate may have for default serialization. */
+inline constexpr auto g_maxMemberCount = 10ull;
+
+/** @brief A simple interface for reading and writing binary data. */
+class BinaryBuffer
+{
+    public:
+        /** @brief Construct a new BinaryBuffer. */
+        explicit BinaryBuffer(size_t sizeHint = 1024)
+        {
+            m_data.reserve(sizeHint);
+        }
+
+        /** @brief Construct a BinaryBuffer from a vector of bytes. */
+        explicit BinaryBuffer(std::vector<uint8_t> data)
+            : m_data{std::move(data)}, m_writePos{m_data.size()}
+        {
+        }
+
+        /** @brief Read bytes from the buffer. */
+        void Read(void* out, size_t nBytes)
+        {
+            AssertRead(nBytes);
+            std::memcpy(out, m_data.data() + m_readPos, nBytes);
+            m_readPos += nBytes;
+        }
+
+        /** @brief Read a trivially copyable object from the buffer. */
+        template<type::TriviallyCopyable T>
+        void Read(T& out)
+        {
+            Read(static_cast<void*>(&out), sizeof(T));
+        }
+
+        /** @brief Write bytes to the buffer. */
+        void Write(const void* in, size_t nBytes)
+        {
+            EnsureCapacity(nBytes);
+            std::memcpy(m_data.data() + m_writePos, in, nBytes);
+            m_writePos += nBytes;
+        }
+
+        /** @brief Write a trivially copyable object to the buffer. */
+        template<type::TriviallyCopyable T>
+        void Write(const T& in)
+        {
+            Write(static_cast<const void*>(&in), sizeof(T));
+        }
+
+        /** @brief Get the number of bytes left to read from the buffer. */
+        auto AvailableReadBytes() const noexcept -> size_t
+        {
+            return m_writePos - m_readPos;
+        }
+
+        /** @brief Get the number of bytes available for writing without requiring a buffer resize. */
+        auto AvailableWriteCapacity() const noexcept -> size_t
+        {
+            return m_data.capacity() - m_writePos;
+        }
+
+        /** @brief Change the capacity of the buffer.
+         *  @note Read and write positions will be updated if nBytes is less than Size(). */
+        void Resize(size_t nBytes)
+        {
+            m_data.resize(nBytes, 0);
+            m_readPos = std::min(m_readPos, nBytes);
+            m_writePos = std::min(m_writePos, nBytes);
+        }
+
+        /** @brief Clear the buffer and reset the read and write positions. */
+        void Clear() noexcept
+        {
+            m_data.clear();
+            m_readPos = 0;
+            m_writePos = 0;
+        }
+
+        /** @brief Release ownership of the underlying binary data. */
+        auto ReleaseBuffer() -> std::vector<uint8_t>
+        {
+            auto buffer = std::move(m_data);
```

Comment:
This should first resize the buffer to m_writePos to get rid of garbage bits past it.